### ISSUE TITLE
Visualize result as confidence map

### ISSUE THREAD

**Issue by trminh89** [alexgkendall/SegNet-Tutorial#10]
Hi @alexgkendall ,

I followed this tutorial (http://nbviewer.ipython.org/github/BVLC/caffe/blob/master/examples/net_surgery.ipynb) and it has a nice visualization of confidence map for classification results.

Can we do the same thing with segnet? If you have time, it would be great if you can show this kind of demo.

Thank so much!

**Comment by alexgkendall** [comment]
Sailency maps like this become a little more complicated with SegNet's dense segmentation.

On a related note, you may be interested in a Bayesian extension of SegNet which produces per class uncertainty maps:
http://arxiv.org/abs/1511.02680

**Comment by trminh89** [comment]
Thanks @alexgkendall ! I'm reading your paper.

Do you have the plan to release Bayesian demo? I would love to see the uncertainty map of each class. I hope you could release a simple demo version (e.g. basic_bayesian_), which may not have the accuracy as the state-of-the-art since I mainly want to know how to visualize the confidence map for this problem (I'm very new to python and caffe).

Many thanks!

**Comment by etienne87** [comment]
Hello there! You can do the exact same thing. For example, in segnet_basic, just forward the net until "prob" layer, then select the channel of interest. It will output confidence map for each pixel.

**Comment by alexgkendall** [comment]
I'm in the process of putting together the tutorial and code for Bayesian SegNet. I'll hopefully release it in a week or so!

@etienne87 what you propose is to display the softmax output as probability. Softmax produces a relative probability between the classes. This is quite different to Bayesian SegNet which produces a global measure of model uncertainty. Have a look at the paper for a further explanation.

Thanks!
Alex

**Comment by etienne87** [comment]
@alexgkendall : will do! (@trminh89  was pointing to caffe tutorial which also plots softmax outputs "tiger cat" channel)

**Comment by arushk1** [comment]
@etienne87 Hey, how did you generate the bounding boxes for the cars? And labels for the other classes?

**Comment by trminh89** [comment]
@alexgkendall : Thanks! I'm waiting for it!
@etienne87 : Did you use segnet for your demo images? I try to reshape the input image and forward it to the net to get  out = net.forward()['prob'][0] output. However, I check all out[label] and they don't show clearly results as in your image. Do we need to scale or do something else with the output? Can you please write you code in more details? Thanks!

**Comment by etienne87** [comment]
@arushk1 : nothing fancy really, just opencv's region growing from argmax of softmax pseudo-probabilities. 

@trminh89  : this is just hmi really, i just did a small cython module to quickly convert single channel class mask  to rgb colors & resized the result a little bit. i'll try to put this small code in a fork.

**Comment by trminh89** [comment]
@etienne87 : Thanks! I run your code and got the following error:
    out = net.forward(data=dd)['prob'][0]    
  File "/home/workspace/segnet/caffe-segnet/python/caffe/pycaffe.py", line 87, in _Net_forward
    raise Exception('Input blob arguments do not match net inputs.')
Exception: Input blob arguments do not match net inputs.

I try to transpose 'im', add new axis and replace you code with:
im = np.transpose(im, (2, 0, 1))
im = im[np.newaxis,:,:,:]
net.blobs['data'].data[...] = im
out = net.forward()['prob'][0]    

Is that ok? It can run now, however, it still difficult to see the clear result as yours. Maybe it because of the accuracy of the net? (your script said my accuracy is only 35% for 1 single image, while if I use @alexgkendall Matlab script, the global acc. is 65%). 

@alexgkendall : Could you please upload the weight model of segnet_basic or segnet for CamVid? I don't have a good graphic card to train now. Thanks!

**Comment by etienne87** [comment]
@trminh89 : I can upload the weights in the fork if you like, but I don't guarantee any performance. For the problem of input_dim, i think it's probably due to the fact that you did not change the input size in the inference model right?  just change the batch size to 1. it should work out.

**Comment by trminh89** [comment]
@etienne87 : Yes, please upload the weights that you used. I'm checking the batch size now! Thanks!!!

**Comment by etienne87** [comment]
@trminh89 : here you go
https://drive.google.com/file/d/0B6HAeMEjid-1V0pCUFdRNERTX28/view?usp=sharing

**Comment by trminh89** [comment]
@etienne87 : Thank you so much! I manage to get the same result as you did! Thanks again!

**Comment by alexgkendall** [comment]
Hey, the Bayesian SegNet tutorial is here if you're interested!
http://mi.eng.cam.ac.uk/projects/segnet/tutorial.html#bayes_segnet
